src/sage/matroids/matroid.pyx

```diff
@@ -7622,15 +7722,63 @@ cdef class Matroid(SageObject):
         R = ZZ['x, y']
         x, y = R._first_ngens(2)
         T = R(0)
-        for B in self.bases():
+        for B in self.bases_iterator():
             T += x ** len(self._internal(B)) * y ** len(self._external(B))
         if a is not None and b is not None:
             T = T(a, b)
         return T
 
+    cpdef characteristic_polynomial(self, l=None) noexcept:
+        r"""
+        Return the characteristic polynomial of the matroid.
+
+        The *characteristic polynomial* of a matroid `M` is the polynomial
+
+        .. MATH::
+
+            \chi_M(\lambda) = \sum_{S \subseteq E} (-1)^{|S|}\lambda^{r(E)-r(S)},
+
+        where `E` is the groundset and `r` is the matroid's rank function. The
+        characteristic polynomial is also equal to
+        `\sum_{i = 0}^r w_i\lambda^{r-i}`, where `\{w_i\}_{i=0}^r` are the
+        Whitney numbers of the first kind.
+
+        INPUT:
+
+        - ``l`` -- a variable or numerical argument (optional)
+
+        OUTPUT: the characteristic polynomial, `\chi_M(\lambda)`, where
+        `\lambda` is substituted with any value provided as input.
+
+        EXAMPLES::
+
+            sage: M = matroids.CompleteGraphic(5)
+            sage: M.characteristic_polynomial()
+            l^4 - 10*l^3 + 35*l^2 - 50*l + 24
+            sage: M.characteristic_polynomial().factor()
+            (l - 4) * (l - 3) * (l - 2) * (l - 1)
+            sage: M.characteristic_polynomial(5)
+            24
+
+        .. SEEALSO::
+
+            :meth:`whitney_numbers() <sage.matroids.matroid.Matroid.whitney_numbers>`
+        """
+        val = l
+        R = ZZ['l']
+        l = R._first_ngens(1)[0]
+        chi = R(0)
+        cdef int r = self.rank()
+        cdef list w = self.whitney_numbers()
+        for i in range(len(w)):
+            chi += w[i] * l**(r-i)
+        if val is not None:
+            return chi(val)
+        return chi
```

Comment:
```suggestion
        chi = self._generic_char_poly()
        if la is not None:
            return chi(la)
        return chi

    @cached_method
    def _generic_char_poly(self):
        R = ZZ['l']
        cdef list w = self.whitney_numbers()
        w.reverse()
        return R(w)
```
I would cache the generic case and then call as appropriate. Also, let the poly ring handle the list itself; it likely can construct it faster.

Reply:
This wouldn't compile so I did something in between.